### PR TITLE
Re-parse vault configuration on every configure() call

### PR DESCRIPTION
**Pull request by fishi0x01**
### Your checklist for this pull request

🚨 Please review the [guidelines for contributing](../blob/master/docs/CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!

- [x] Please describe what you did

- [ ] Link to relevant GitHub issues or pull requests

- [ ] Link to relevant [Jenkins JIRA issues](https://issues.jenkins-ci.org)

- [x] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### Description

Currently, `VaultSecretSource` does not re-read the `CASC_VAULT_*` configuration parameters on subsequent `configure()` calls. In order to change the vault configuration, Jenkins needs to be restarted. 

In this PR the `CASC_VAULT_*` parameters are read on every `configure()` call. If changes are detected on VaultAuthenticators, then a new authentication is triggered. 

The current test cases already have subsequent `VaultSecretSource.init()` calls.

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java [hunk 1]
```diff
@@ -49,10 +49,13 @@
     private Map<String, String> secrets = new HashMap<>();
     private Vault vault;
     private VaultConfig vaultConfig;
-    private VaultAuthenticator vaultAuthenticator;
+    private VaultAuthenticator currentlyUsedAuthenticator;
+    private VaultAuthenticator configuredAuthenticator;
```
Comment (timja) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: does this need to be a field? looks like its only used inside the configureVault method,
I think you could inline it and undo the field rename above?

Comment (fishi0x01) [reply plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: You are right, but I wasn't sure how to do it. I am using lambda expressions https://github.com/jenkinsci/configuration-as-code-plugin/pull/810/files#diff-f7f1df35659ecbdcd0652b07007b223fR111 

I am getting the following if I declare it in a local function context: 
```
Variable used in lambda expression should be final or effectively final
```

I could remove the lambda functions, but that would be replaced by lots of `!= null` checks, so I thought declaring `configuredAuthenticator` in class context would be easy way out.

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultAuthenticator.java [hunk 2]
```diff
@@ -6,4 +6,5 @@
 
 public interface VaultAuthenticator {
     void authenticate(Vault vault, VaultConfig config) throws VaultException;
+    boolean equalsAuthenticator(VaultAuthenticator authenticator);
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultAuthenticator.java]: use a default implementation :smile:

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultSingleTokenAuthenticator.java [hunk 3]
```diff
@@ -15,4 +15,8 @@ public void authenticate(Vault vault, VaultConfig config) throws VaultException
         // No special mechanism - token already exists
         config.token(token).build();
     }
+
+    public boolean equalsAuthenticator(VaultAuthenticator authenticator) {
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultSingleTokenAuthenticator.java]: seems like the default

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultAppRoleAuthenticator.java [hunk 4]
```diff
@@ -21,10 +23,20 @@ public VaultAppRoleAuthenticator(String approle, String approleSecret) {
...
+            config.token(currentAuthToken).build();
             LOGGER.log(Level.FINE, "Login to Vault using AppRole/SecretID successful");
             getTTLExpiryOfCurrentToken(vault);
+        } else {
+            // make sure current auth token is set in config
+            config.token(currentAuthToken).build();
         }
     }
+
+    public String getAttributeHash() {
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultAppRoleAuthenticator.java]: Should this method not be declared on the interface?

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultSingleTokenAuthenticator.java (hunk 3 above)
Comment (casz) [reply plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultSingleTokenAuthenticator.java]: This is unused now?

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultAuthenticatorWithExpiration.java [hunk 5]
```diff
@@ -45,4 +45,13 @@ public void getTTLExpiryOfCurrentToken(Vault vault) {
         tokenExpiration = Calendar.getInstance();
         tokenExpiration.add(Calendar.SECOND, tokenTTL);
     }
+
+    public boolean equalsAuthenticator(VaultAuthenticator o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        VaultAuthenticatorWithExpiration authenticator = (VaultAuthenticatorWithExpiration) o;
+        return getAttributeHash().equals(authenticator.getAttributeHash());
```
[the remark this answers is not in the document]
Comment (fishi0x01) [reply plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/vault/VaultAuthenticatorWithExpiration.java]: Yup you are right. Somehow the CI was complaining when I override `equals()`, which is why I used a custom one then instead. But I think my mistake was that I did not override `hashCode()` too. Thx for pointing out :bowing_man:

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java [hunk 6]
```diff
@@ -105,22 +108,28 @@ private void determineAuthenticator(Properties prop) {
...
-                        approleSecret -> vaultAuthenticator = new VaultAppRoleAuthenticator(approle, approleSecret)
+                        approleSecret -> configuredAuthenticator = new VaultAppRoleAuthenticator(approle, approleSecret)
                 )
         );
 
-        if (vaultAuthenticator == null) {
+        // Overwrite current authenticator only if there was a change, because we do not want to loose current auth token
+        if (configuredAuthenticator != null
+                && !configuredAuthenticator.equals(currentlyUsedAuthenticator)) {
+            currentlyUsedAuthenticator = configuredAuthenticator;
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: Move to a method, avoiding the need for two VaultAuthenticator fields and cleans up lambdas too. You could implement the new `YamlSource.of` Method and reduce it even more to an actual method reference.

https://github.com/jenkinsci/configuration-as-code-plugin/blob/f16aebb28d5c29aeb9d9582d05e8772d92708d26/plugin/src/main/java/io/jenkins/plugins/casc/yaml/YamlSource.java#L25-L48

Comment (casz) [reply plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: I started implementing, cause I wanted to see the code :smile:

**Comment by casz** [comment]
:tada: that cleaned up a lot of code.

I moved to `properties` to a private field since it became a burden to pass along everywhere especially with lambdas

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java [hunk 7]
```diff
@@ -96,35 +96,48 @@ public VaultSecretSource() {
...
+        // Overwrite current authenticator only if there was a change, because we do not want to loose current auth token
+        if (vaultAuthenticator != null
+            && !vaultAuthenticator.equals(this.vaultAuthenticator)) {
+            this.vaultAuthenticator = vaultAuthenticator;
+        }
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static <T, U> void allPresent(Optional<T> o1, Optional<U> o2, BiConsumer<T, U> consumer) {
+        o1.ifPresent(t -> o2.ifPresent(u -> consumer.accept(t, u)));
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: @dgarzon wonder if there is more io.vavr way, I tried but no dice.

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java (hunk 6 above)
Comment (casz) [reply plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: @fishi0x01 What do you think of my additions? 😅

Comment (fishi0x01) [reply plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: Yes looks awesome! Thank you :)

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java [hunk 8]
```diff
@@ -5,10 +5,8 @@
 import com.bettercloud.vault.VaultException;
 import hudson.Extension;
 import io.jenkins.plugins.casc.SecretSource;
-import io.jenkins.plugins.casc.impl.secrets.vault.VaultAppRoleAuthenticator;
 import io.jenkins.plugins.casc.impl.secrets.vault.VaultAuthenticator;
-import io.jenkins.plugins.casc.impl.secrets.vault.VaultSingleTokenAuthenticator;
-import io.jenkins.plugins.casc.impl.secrets.vault.VaultUserPassAuthenticator;
+import java.util.function.BiConsumer;
```
Comment (casz) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: The import order, should be fixed :|

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java (hunk 7 above)
Comment (casz) [reply plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java]: @dgarzon I would see be interested in an answer on io.vavr way of doing this 😓

**Pull request merged**